compiler-core/src/type_/expression.rs

```diff
@@ -704,6 +712,21 @@ impl<'a, 'b, 'c> ExprTyper<'a, 'b, 'c> {
                 .map_err(|e| e.case_clause_mismatch().into_error(typed_clause.location()))?;
             typed_clauses.push(typed_clause);
         }
+
+        for i in 0..subjects_count {
+            let t = collapse_links(subject_types[i].clone());
+            let mut patterns: Vec<Pattern<PatternConstructor, Arc<Type>>> = Vec::new();
+            for c in &typed_clauses {
+                patterns.push(c.pattern[i].clone());
+                for a in &c.alternative_patterns {
+                    patterns.push(a[i].clone());
+                }
+            }
+            if !self.environment.exhaustive(patterns, t) {
+                return Err(Error::NotExhaustivePatternMatch { location });
+            }
+        }
```

Comment:
Just realized this doesn't fully work, e.g. doesn't warn for situations like
```
  case b1, b2 {
    True, False -> Nil
    False, True -> Nil
  }
```

I could do some combinatorics to check if all possibilities are covered but maybe for this simple limited solution only do the check if there's one subject? 🤔

Reply:
Re-watched SPJ's talk linked in the ticket, they store a "refinement types" style (set theoretic) representation of possible values, and when iterating through clauses they refine/narrow down the sets by what matches given clause. That's pretty wild.

Reply:
I'd like to implement [this algorithm](https://github.com/gleam-lang/gleam/issues/34#issuecomment-896821619) later which will handle this. The Rust reference should help.

For now it's OK to have an imperfect solution.

Reply:
I looked at the linked algorithm, which confirmed my suspicions that what I did here for multiple subjects is not that useful. I'll replace this code with a simpler check that takes effect only when there's one subject. I think it'll be less confusing to the user